docs/templated_annotations.md

```diff
@@ -0,0 +1,217 @@
+# Templating
+
+## `@template`
+
+The `@template` tag allows classes and functions to implement type parameter-like functionality found in many other languages.
+
+While `@template` tag order matters (i.e. for key-value pair extending), names don't matter outside the scope of the class or function in which they're declared.
+
+As a very simple example, this function returns whatever is passed in:
+
+```php
+/**
+ * @template T
+ * @psalm-param T $t
+ * @return T
+ */
+function mirror($t) {
+    return $t;
+}
+
+$a = 5;
+$b = mirror(5); // Psalm knows the result is an int
```

Comment:
```suggestion
$b = mirror(5); // Psalm knows the result is an int

$c = "foo";
$d = mirror($c); // Psalm knows the result is string
```

Reply:
can't quite tell from github's preview- is the suggestion to remove line 21?

Reply:
no, to add couple of more lines, showing the method return type being polymorphic based on argument type.

Reply:
ah :D